File: plugins/battery.30s.go
// +build ignore

package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	level, charging := getBatteryStatus()

	if level < 0 {
		fmt.Println("\U0001F50C N/A")
		return
	}

	// Determine icon and color based on level and charging state
	var icon, color string
	if charging {
		icon = "\u26A1" // lightning bolt
		color = "blue"
	} else {
		switch {
		case level <= 10:
			icon = "\U0001FAAB" // empty battery
			color = "red"
		case level <= 25:
			icon = "\U0001F50B" // battery
			color = "orange"
		case level <= 50:
			icon = "\U0001F50B"
			color = "yellow"
		default:
			icon = "\U0001F50B"
			color = "green"
		}
	}

	fmt.Printf("%s %d%% | color=%s\n", icon, level, color)
	fmt.Println("---")
	fmt.Printf("Battery Level: %d%%\n", level)
	if charging {
		fmt.Println("Status: Charging")
	} else {
		fmt.Println("Status: Discharging")
	}
	fmt.Println("---")
	fmt.Println("Refresh | refresh=true")
}

func getBatteryStatus() (int, bool) {
	switch runtime.GOOS {
	case "linux":
		return getLinuxBattery()
	case "darwin":
		return getMacOSBattery()
	default:
		return -1, false
	}
}

func getLinuxBattery() (int, bool) {
	// Find battery in /sys/class/power_supply/
	matches, _ := filepath.Glob("/sys/class/power_supply/BAT*")
	if len(matches) == 0 {
		return -1, false
	}

	batPath := matches[0]

	// Read capacity
	capacityData, err := os.ReadFile(filepath.Join(batPath, "capacity"))
	if err != nil {
		return -1, false
	}
	level, _ := strconv.Atoi(strings.TrimSpace(string(capacityData)))

	// Read status
	statusData, err := os.ReadFile(filepath.Join(batPath, "status"))
	charging := false
	if err == nil {
		status := strings.TrimSpace(string(statusData))
		charging = status == "Charging" || status == "Full"
	}

	return level, charging
}

func getMacOSBattery() (int, bool) {
	// On macOS, we would use pmset or IOKit
	// For simplicity, return -1 (not available)
	return -1, false
}

File: plugins/battery.30s.rs
use std::process::Command;
use std::str;
use std::str::FromStr;

fn main() {
    let (level, charging) = get_battery_status_crossbar();

    if level < 0 {
        println!("\u{1F50C} N/A");
        return;
    }

    // Determine icon and color based on level and charging state
    let (icon, color) = if charging {
        ("\u{26A1}", "blue") // lightning bolt
    } else {
        match level {
            0..=10 => ("\u{1FAAB}", "red"), // empty battery
            11..=25 => ("\u{1F50B}", "orange"), // battery
            26..=50 => ("\u{1F50B}", "yellow"),
            _ => ("\u{1F50B}", "green"),
        }
    };

    println!("{} {}% | color={}", icon, level, color);
    println!("---");
    println!("Battery Level: {}%", level);
    if charging {
        println!("Status: Charging");
    } else {
        println!("Status: Discharging");
    }
    println!("---");
    println!("Refresh | refresh=true");
}

fn get_battery_status_crossbar() -> (i32, bool) {
    let output = Command::new("crossbar")
        .arg("battery")
        .output();

    match output {
        Ok(cmd_output) => {
            let battery_info = str::from_utf8(&cmd_output.stdout).unwrap_or("").trim();
            // Example: "87% ⚡" or "50%"

            let mut level_str = String::new();
            for c in battery_info.chars() {
                if c.is_ascii_digit() {
                    level_str.push(c);
                } else if !level_str.is_empty() {
                    break;
                }
            }

            let level = i32::from_str(&level_str).unwrap_or(-1);
            let charging = battery_info.contains("⚡");

            (level, charging)
        }
        Err(_) => (-1, false),
    }
}
